### DISCUSSION TITLE
Questions regarding new exponential envelopes

### DISCUSSION THREAD

**Discussion by cdw2000** [ElectroTechnique/electrotechnique.github.io#58] [Software Issues]
The new exponential envelope generator is great, but I have a few questions regarding the selectable curves for Attack.   The range is selectable from -8 to +8.   Does that mean 0 is effectively linear?   It looks like the curve is adjustable from exponential to logarithmic.    Are the negative values for the logarithmic curves or are the positive values?   The documentation isn't too clear.

Also, what values represent a "typical" analog RC circuit style curve?

Reference this link for why I'm asking:  https://learningmodular.com/linear-versus-exponential/

Thanks, and once again, great work!

**Comment by Electrotechnique** [comment]
Hello Chris,

Yes, 0 appears to be linear but there's also a linear setting (Lin). Looking at the shapes produced they are the same. I believe Lin bypasses the code that produces the exponential shape, looking at the code. 

Values -8 to -1 produce an RC style attack 1 - e(-x), where the shape mimics a capacitor charging to a peak value.
This, for anyone who's not into electronics:
![external-content duckduckgo com](https://user-images.githubusercontent.com/50934180/123267787-5a039080-d538-11eb-9257-b7f38b395107.jpg)

**Reply by vipear** [reply]
All of the curves are exponential. Even the 0 case can be thought of as exponential in the limiting case. The general form of the attack equation is K*(1-exp(a*t)). For negative values of 'a' K is positive and the curve is like a charging capacitor. For positive 'a' K is chosen to be negative and the curve starts at zero and rapidly increases in which there is no passive circuit equivalent. The constant K and the exponent 'a' are chosen to hit the maximum value in a specified amount of time. If 'a' approaches zero then K has to approach infinity in which case the Taylor series approximation of the equation reduces to K*a*t where K*a is constant and the curve is linear. The constant K cannot be solved directly with 'a' equal to zero with finite precision math using the same equation. I believe this is called a removable singularity. In that case I treat it as simply a linear segment. The link given above I believe is incorrect in the statement that most envelope generators use a logarithmic attack. Most analog envelope generators charge a capacitor which is an exponential curve. You can change sign of an exponential or its exponent and add or subtract constant terms from it but it is still exponential. Some generators may have logarithmic segments but it is much more complicated to do with analog electronics. It is easier to do it digitally. Confusion as to being exponential or logarithmic seems to come from incorrectly interpreting the curvature. A logarithmic curve would require swapping the time and amplitude axis (along with arbitrary sign and added or subtracted constant).

**Reply by vipear** [reply]
The general equation used for calculating the attack for any of the curves (except n=0) is A(t)=(1-exp(t*n/Ta))/(1-exp(n)). For any n not equal to zero results in a series of convex (n<0) or concave (n>0) curves that all start at zero at time t=0 and end at 1 at t equal to the attack time Ta. The Taylor expansion of the equation about the variable n as n approaches zero is t/Ta, a linear segment with the same endpoints. However, plugging 0 into the equation directly results in the undefined 0/0. Like the sinc function sin(x)/x it has a value as x approaches zero. Here it is treated as a special case.
